### ISSUE TITLE
Problem joining tables where string ID matches in MySQL but not in Prisma

### ISSUE THREAD

**Issue by stuartkeith** [prisma/prisma#15073]
### Bug description

If I have a primary key on one table and another column referencing that primary key, and one primary key has (for example) a [no-break-space](https://unicode-table.com/en/00A0/) character in it, MySQL will treat those keys as equivalent (when joining or enforcing constraints), but a Prisma query attempting to join them fails with "Inconsistent query result: Field X is required to return data, got `null` instead."

This is using the default `utf8mb4_unicode_ci` collation. I assume this issue could arise for other characters that MySQL collates.

### How to reproduce

I've made a repo showing this issue here: https://github.com/stuartkeith/prisma-issue-0

There's a Pet table with a string primary key, and an Owner table with a reference to that primary key:

```
model Pet {
  id    String  @id
  Owner Owner[]
}

model Owner {
  id Int @id @default(autoincrement())

  pet   Pet    @relation(fields: [petId], references: [id])
  petId String
}
```

Insert a Pet with an ID containing a no-break-space character, and then an Owner with a reference to that ID without the no-break-space character:

```
  await prisma.pet.create({
    data: {
      id: "dog" + String.fromCharCode(parseInt('00A0', 16)) + "example"
    }
  });

  await prisma.owner.create({
    data: {
      petId: "dog example"
    }
  });
```

MySQL allows the owner to be created even though the strings don't 100% match, and doesn't restrict joining on those IDs:

```
await prisma.$queryRaw`SELECT o.petId as ownerPetId, p.id as petId FROM Owner o INNER JOIN Pet p ON o.petId = p.id`
> [ { ownerPetId: 'dog example', petId: 'dog example' } ]
```

This same query in Prisma doesn't work, there is a "Inconsistent query result: Field pet is required to return data, got `null` instead", I assume because Prisma is a lot stricter about matching strings when finding IDs in the result.

```
  // throws
  const result = await prisma.owner.findFirst({
    where: {
      petId: "dog example"
    },
    select: {
      id: true,
      pet: {
        select: {
          id: true
        }
      }
    }
  })
```

In this case I fixed this by removing the character, it wasn't meant to be there, but it was a tricky scenario to debug considering the character is invisible and MySQL itself never complained about the different characters.

### Expected behavior

I expect Prisma's method of matching strings to be consistent with MySQL.

### Prisma information

See https://github.com/stuartkeith/prisma-issue-0

```
model Pet {
  id    String  @id
  Owner Owner[]
}

model Owner {
  id Int @id @default(autoincrement())

  pet   Pet    @relation(fields: [petId], references: [id])
  petId String
}
```

```
  await prisma.pet.create({
    data: {
      // append NO-BREAK-SPACE
      // https://www.fileformat.info/info/unicode/char/00a0/index.htm
      id: "dog" + String.fromCharCode(parseInt('00A0', 16)) + "example"
    }
  });

  // can successully refer to the key without any space
  await prisma.owner.create({
    data: {
      petId: "dog example"
    }
  });

  // works
  const rawResult = await prisma.$queryRaw`SELECT o.petId as ownerPetId, p.id as petId FROM Owner o INNER JOIN Pet p ON o.petId = p.id`;

  // throws
  const result = await prisma.owner.findFirst({
    where: {
      petId: "dog example"
    },
    select: {
      id: true,
      pet: {
        select: {
          id: true
        }
      }
    }
  })
```

### Environment & setup

- OS: MaC OS
- Database: MySQL
- Node.js version: v16.13.2

### Prisma Version

```
prisma                  : 4.2.1
@prisma/client          : 4.2.1
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine 2920a97877e12e055c1333079b8d19cee7f33826 (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli 2920a97877e12e055c1333079b8d19cee7f33826 (at node_modules/@prisma/engines/migration-engine-darwin-arm64)
Introspection Engine    : introspection-core 2920a97877e12e055c1333079b8d19cee7f33826 (at node_modules/@prisma/engines/introspection-engine-darwin-arm64)
Format Binary           : prisma-fmt 2920a97877e12e055c1333079b8d19cee7f33826 (at node_modules/@prisma/engines/prisma-fmt-darwin-arm64)
Default Engines Hash    : 2920a97877e12e055c1333079b8d19cee7f33826
Studio                  : 0.469.0
```

**Comment by Jolg42** [comment]
Can reproduce the ` prisma.owner.findFirst` error on 4.2.1 and 4.3.1
```
"Inconsistent query result: Field pet is required to return data, got `null` instead"
```

**Comment by LinusU** [comment]
I'm currently trying to figure out what the best practices are for using string ids in MySQL. We are migrating from another system that used fixed-length random strings as ids, so we'll use that here as well. It's important for us that a case-sensitive comparison is made between the ids since they use both lower and upper cased letters.

My current attempt looks something like this:

```prisma
model Foo {
  id String @id @db.Char(10)
}
```

I then modify the migration manually like so:

```diff
  CREATE TABLE `Foo` (
-     `id` CHAR(10) NOT NULL,
+     `id` CHAR(10) CHARACTER SET binary NOT NULL,
```

**Comment by janpio** [comment]
@LinusU Please open a new discussion, your request does not actually seem related to the bug report here. (If this bug report dds useful context, of course feel free to link to it). Thanks.

**Comment by LinusU** [comment]
@janpio sorry, might have been a bit unclear, but what I posted should be a workaround for the bug described in the OP since it should make MySQL compare the strings byte-for-byte instead of using locale aware rules. That should then match the logic that Prisma uses (which I assume is `===` in JavaScript)

**Comment by aqrln** [comment]
Hi @stuartkeith! We have added support for MySQL 8.0.14 and newer to the `relationJoins` preview feature in Prisma 5.10.0, which we expect to resolve this issue thanks to the database comparing the IDs itself when joining the data.

Does enabling the `relationJoins` preview feature fix this problem for you?

```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["relationJoins"]
}
```

Thank you in advance for following up 🙏

**Comment by stuartkeith** [comment]
Hi @aqrln, thanks for getting back to me. I can confirm that using `relationJoins` fixes this issue.

I pushed a branch to my repo (https://github.com/stuartkeith/prisma-issue-0/compare/fixed?expand=1) that shows updating to 5.10.0 with `relationJoins` fixes the issue.

Thanks for fixing this!

**Comment by aqrln** [comment]
Thanks a lot of the confirmation @stuartkeith!

Heads up that the `relationJoins` preview feature currently has some limitations, and it will fall back to the old query strategy if a query contains `distinct` (on MySQL; in some cases on Postgres as well) or `cursor` (always). This means those queries would still be affected by this issue. We are actively working on closing the gaps: for example, `distinct` support with `relationJoins` on MySQL is going to be shipped in the next release.

If you have any feedback regarding the preview feature, don't hesitate to drop a comment in https://github.com/prisma/prisma/discussions/22288. We're especially interested in performance differences, but also everything else as well! Of course also please let us know if you run into any problems. Thank you!
